### PR TITLE
Web: Workaround Emscripten 3.1.42+ LTO regression

### LINKED ISSUE #80010: Compiling Godot web export template with LTO fails with Emscripten 3.1.42+
### Godot version

3.5.2

### System information
AMD EPYC 7B13; 32 GB RAM
Ubuntu 22.04.2 LTS; Python 3.11.1; Emscripten 3.1.44; clang 17; Scons 4.0.1

### Issue description

Compiling with LTO gives this error:
```
[Initial build] Compiling ==> core/bind/core_bind.cpp
[Initial build] Creating 'bin/godot.javascript.opt.engine.js'
[Initial build] Install file: "misc/dist/html/offline-export.html" as "bin/.javascript_zip/godot.offline.html"
[Initial build] Install file: "misc/dist/html/service-worker.js" as "bin/.javascript_zip/godot.service.worker.js"
[Initial build] progress_finish(["progress_finish"], [])
[Initial build] Linking Static Library ==> core/libcore.javascript.opt.a
Ranlib Library         ==> core/libcore.javascript.opt.a
[Initial build] Linking Program        ==> bin/godot.javascript.opt.js
cache:INFO: generating system asset: symbol_lists/df45337bb9df6876919ba6b0bb4825c9a26cbd2f.json... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/symbol_lists/df45337bb9df6876919ba6b0bb4825c9a26cbd2f.json" for subsequent builds)
cache:INFO:  - ok
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libGL-webgl2-ofb.a... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libGL-webgl2-ofb.a" for subsequent builds)
system_libs:INFO: compiled 4 inputs
cache:INFO:  - ok
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libal.a... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libal.a" for subsequent builds)
system_libs:INFO: compiled 1 inputs
cache:INFO:  - ok
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libhtml5.a... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libhtml5.a" for subsequent builds)
system_libs:INFO: compiled 3 inputs
cache:INFO:  - ok
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libstubs.a... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libstubs.a" for subsequent builds)
system_libs:INFO: compiled 2 inputs
cache:INFO:  - ok
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libc.a... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libc.a" for subsequent builds)
system_libs:INFO: compiled 1023 inputs
cache:INFO:  - ok
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libdlmalloc.a... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libdlmalloc.a" for subsequent builds)
system_libs:INFO: compiled 1 inputs
cache:INFO:  - ok
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libcompiler_rt.a... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libcompiler_rt.a" for subsequent builds)
system_libs:INFO: compiled 175 inputs
cache:INFO:  - ok
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libc++-noexcept.a... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libc++-noexcept.a" for subsequent builds)
system_libs:INFO: compiled 49 inputs
cache:INFO:  - ok
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libc++abi-noexcept.a... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libc++abi-noexcept.a" for subsequent builds)
system_libs:INFO: compiled 16 inputs
cache:INFO:  - ok
cache:INFO: generating system library: sysroot/lib/wasm32-emscripten/lto/libsockets.a... (this will be cached in "/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libsockets.a" for subsequent builds)
system_libs:INFO: compiled 21 inputs
cache:INFO:  - ok
wasm-ld: error: /workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto/libc.a(scalbnf.o): attempt to add bitcode file after LTO.
em++: error: '/workspace/web/emsdk/upstream/bin/wasm-ld -o bin/godot.javascript.opt.wasm platform/javascript/audio_driver_javascript.javascript.opt.o platform/javascript/godot_webgl2.javascript.opt.o platform/javascript/http_client_javascript.javascript.opt.o platform/javascript/javascript_singleton.javascript.opt.o platform/javascript/javascript_main.javascript.opt.o platform/javascript/os_javascript.javascript.opt.o platform/javascript/api/javascript_tools_editor_plugin.javascript.opt.o platform/javascript/javascript_runtime.javascript.opt.o main/libmain.javascript.opt.a modules/libmodules.javascript.opt.a platform/libplatform.javascript.opt.a drivers/libdrivers.javascript.opt.a scene/libscene.javascript.opt.a servers/libservers.javascript.opt.a core/libcore.javascript.opt.a modules/freetype/libfreetype_builtin.javascript.opt.a -L/workspace/web/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten/lto -lGL-webgl2-ofb -lal -lhtml5 -lstubs -lc -ldlmalloc -lcompiler_rt -lc++-noexcept -lc++abi-noexcept -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr /tmp/tmpk0a0yca9libemscripten_js_symbols.so --strip-debug --export-if-defined=main --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__start_em_lib_deps --export-if-defined=__stop_em_lib_deps --export-if-defined=__start_em_js --export-if-defined=__stop_em_js --export-if-defined=__main_argc_argv --export-if-defined=fflush --export=stackSave --export=stackRestore --export=stackAlloc --export=__errno_location --export=__get_temp_ret --export=__set_temp_ret --export=__funcs_on_exit --export=__wasm_call_ctors --export-table -z stack-size=65536 --initial-memory=33554432 --no-entry --max-memory=2147483648 --global-base=1024' failed (returned 1)
scons: *** [bin/godot.javascript.opt.js] Error 1
scons: building terminated because of errors
```

### Steps to reproduce

1. Install the requirements listed on https://docs.godotengine.org/en/3.6/development/compiling/compiling_for_web.html.
2. Run `scons platform=javascript tools=no target=release use_lto=yes`

### I did not have this issue before. I am using a fresh build setup.

### PR DESCRIPTION
**Pull request by akien-mga**
Fixes #80010.

Tested with Emscripten 3.1.45 and `scons p=web target=template_release production=yes verbose=yes`.

### CODE REVIEW

File: .github/workflows/web_builds.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ concurrency:
 
 jobs:
   web-template:
-    runs-on: "ubuntu-20.04"
+    runs-on: "ubuntu-22.04"
```
Comment (akien-mga) [inline .github/workflows/web_builds.yml:19]: This isn't strictly needed, just a drive-by change as Ubuntu 20.04 will be EOL'ed by GH Actions eventually, and we have no reason to stick to the old version for Web builds.

File: platform/web/detect.py [hunk 2]
```diff
@@ -118,6 +118,11 @@ def configure(env: "Environment"):
         else:
             env.Append(CCFLAGS=["-flto"])
             env.Append(LINKFLAGS=["-flto"])
+        # Workaround https://github.com/emscripten-core/emscripten/issues/19781.
+        cc_version = get_compiler_version(env)
+        cc_semver = (int(cc_version["major"]), int(cc_version["minor"]), int(cc_version["patch"]))
+        if cc_semver >= (3, 1, 42):
```
Comment (akien-mga) [inline platform/web/detect.py:124]: I'll refactor this logic in #80719 after this is merged.

**Review by YuriSizov — APPROVED** [review]

**Review by Faless — APPROVED** [review]
LGTM :+1: !

File: platform/web/detect.py (hunk 2 above)
Comment (akien-mga) [reply platform/web/detect.py:124]: Just noticed in the `3.x` branch that we use `tuple(get_compiler_version(env))` there, which is nicer. I'm changing it to that when cherry-picking, and for `master` I'll refactor later.

**Comment by akien-mga** [comment]
Cherry-picked for 3.6.

**Comment by akien-mga** [comment]
Cherry-picked for 3.5.3.

**Comment by Pogie10** [comment]
> Fixes #80010.
> 
> Tested with Emscripten 3.1.45 and `scons p=web target=template_release production=yes verbose=yes`.

FIX UR FACE IDIOT

File: .github/workflows/web_builds.yml (hunk 1 above)
Comment (Pogie10) [reply .github/workflows/web_builds.yml:19]: TALK TO MY HAND 👐

**Comment by Pogie10** [comment]
> LGTM 👍 !

MGA ULUL

**Review by Pogie10 — COMMENTED** [review]
MGA TANGA KAU PAG UUNTUGIN KO MGA ULO NYO

**Comment by YuriSizov** [comment]
Cherry-picked for 4.1.2.
